### PR TITLE
Counter utf-16 characters

### LINKED ISSUE #282: A problem for editing Non-ASCII characters
Hi,

When I edit a file which includes Non-ASCII characters as the comments, string literals, ..etc., vim-lsp goes wrong.

For example:

```typescript
export enum E {
  S = '𐐀',
}
```

When I remove `𐐀` (assigning a empty string to `E.S`) in the code above,

- `Unterminated literal string` error is shown on that line (the line `S` is defined).
- The completion for `E` does not work.

After some investigation, what I've got is:

- The number of `character` in the parameter to the server should be counted on a UTF-16 string representation ([LSP SPEC](https://microsoft.github.io/language-server-protocol/specification#text-documents)).
- `strlen()` is used in vim-lsp for the calculation.

=> vim-lsp sends the wrong `character` parameter to the server when the non-ascii characters (whose code point is more than 0x10000) are edited.

A solution for this problem is, I think, replacing `strlen()` with the function like below:

```VimL
function! s:count_utf16_code_units(str) abort
  let l:len = strchars(a:str)
  let l:i = 0
  let l:cnt = 0

  while l:i < l:len
    let l:chr = strcharpart(a:str, l:i, 1)
    if char2nr(l:chr) > 0x10000
      let l:cnt = l:cnt + 2
    else
      let l:cnt = l:cnt + 1
    endif

    let l:i = l:i + 1
  endwhile

  return l:cnt
endfunction
```

I'm not good at Vim script, so I hope these information helps fixing the problem.

### PR DESCRIPTION
**Pull request by mattn**
Fixes #282

### CODE REVIEW

**Comment by mattn** [comment]
Benchmark

```vim
function! s:count_utf16_code_units(str) abort
  let l:len = strchars(a:str)
  let l:i = 0
  let l:cnt = 0

  while l:i < l:len
    let l:chr = strcharpart(a:str, l:i, 1)
    if char2nr(l:chr) > 0x10000
      let l:cnt = l:cnt + 2
    else
      let l:cnt = l:cnt + 1
    endif

    let l:i = l:i + 1
  endwhile

  return l:cnt
endfunction

function! s:my_count_utf16_code_units(str) abort
  let l:rs = split(a:str, '\zs')
  let l:len = len(l:rs)
  return l:len + count(l:rs, 'char2nr(v:val)>0x10000')
endfunction

function! s:benchmark()
  let s = 'a𐐀b'
  let n = 100000

  let start = reltime()
  for i in range(n)
    call strlen(s)
  endfor
  echo printf('strlen: %f', reltimefloat(reltime(start)))

  let start = reltime()
  for i in range(n)
    call s:count_utf16_code_units(s)
  endfor
  echo printf('s:count_utf16_code_units: %f', reltimefloat(reltime(start)))

  let start = reltime()
  for i in range(n)
    call s:my_count_utf16_code_units(s)
  endfor
  echo printf('s:my_count_utf16_code_units: %f', reltimefloat(reltime(start)))
endfunction

call s:benchmark()
```

```
strlen: 0.229879
s:count_utf16_code_units: 4.050134
s:my_count_utf16_code_units: 1.107014
```

File: autoload/lsp/utils/diff.vim [hunk 1]
```diff
@@ -99,17 +99,23 @@ function! s:Length(lines, start_line, start_char, end_line, end_char)
...
   while l:line < l:adj_end_line
-    let l:result += strlen(a:lines[l:line]) + 1
+    let l:result += s:count_utf16_code_units(a:lines[l:line]) + 1
     let l:line += 1
   endwhile
   let l:result += l:adj_end_char + 1
   return l:result
 endfunction
+
+function! s:count_utf16_code_units(str) abort
```
Comment (natebosch) [inline autoload/lsp/utils/diff.vim]: Does this behave differently from `strchars(str)`?

**Comment by mattn** [comment]
It is no needed that replace all of strlen to this function. This function should be used for only calculate offset.

File: autoload/lsp/utils/diff.vim (hunk 1 above)
Comment (mattn) [reply autoload/lsp/utils/diff.vim]: In LSP spec, `𐐀` is two characters. But strchars return 1. See https://github.com/Microsoft/language-server-protocol/issues/376

**Comment by mattn** [comment]
```vim
function! s:count_utf16_code_units(str) abort
  let l:len = strchars(a:str)
  let l:i = 0
  let l:cnt = 0

  while l:i < l:len
    let l:chr = strcharpart(a:str, l:i, 1)
    if char2nr(l:chr) > 0x10000
      let l:cnt = l:cnt + 2
    else
      let l:cnt = l:cnt + 1
    endif

    let l:i = l:i + 1
  endwhile

  return l:cnt
endfunction

function! s:my_count_utf16_code_units(str) abort
  let l:rs = split(a:str, '\zs')
  let l:len = len(l:rs)
  return l:len + count(l:rs, 'char2nr(v:val)>0x10000')
endfunction

function! s:benchmark(name, n, case)
  let start = reltime()
  for i in range(a:n)
    call strlen(a:case)
  endfor
  echo printf('%s: strlen: %f', a:name, reltimefloat(reltime(start)))

  let start = reltime()
  for i in range(a:n)
    call s:count_utf16_code_units(a:case)
  endfor
  echo printf('%s: s:count_utf16_code_units: %f', a:name, reltimefloat(reltime(start)))

  let start = reltime()
  for i in range(a:n)
    call s:my_count_utf16_code_units(a:case)
  endfor
  echo printf('%s: s:my_count_utf16_code_units: %f', a:name, reltimefloat(reltime(start)))
endfunction

call s:benchmark('short string', 100000, 'a𐐀b')
call s:benchmark('long string', 10000, repeat('a𐐀b', 200))
```
```
short string: strlen: 0.242768
short string: s:count_utf16_code_units: 4.028752
short string: s:my_count_utf16_code_units: 1.097196
long string: strlen: 0.024196
long string: s:count_utf16_code_units: 58.931086
long string: s:my_count_utf16_code_units: 3.932208
```

File: autoload/lsp/utils/diff.vim [hunk 2]
```diff
@@ -99,17 +99,23 @@ function! s:Length(lines, start_line, start_char, end_line, end_char)
...
+  let l:rs = split(a:str, '\zs')
+  let l:len = len(l:rs)
+  return l:len + count(l:rs, 'char2nr(v:val)>0x10000')
```
Comment (k-takata) [inline autoload/lsp/utils/diff.vim]: Shouldn't be `s/>/>=/`?

Comment (mattn) [reply autoload/lsp/utils/diff.vim]: right

**Comment by prabirshrestha** [comment]
There are some interesting approach and comments going on currently at https://github.com/Microsoft/language-server-protocol/issues/376#issuecomment-475894185

**Comment by Avi-D-coder** [comment]
Is vim-lsp currently using UTF-8/bytes, codepoints or grapheme clusters?

**Comment by prabirshrestha** [comment]
There is now webworkers in ducktape vim. Perf heavy things could go there :) https://github.com/bobpepin/vim/issues/1

**Comment by prabirshrestha** [comment]
@mattn Any updates on this? One option I can think of is having this under a feature flag turned off by default.

**Comment by mattn** [comment]
Sorry delay, I'll fix soon.

File: autoload/lsp/utils.vim [hunk 3]
```diff
@@ -185,3 +185,23 @@ function! lsp#utils#echo_with_truncation(msg) abort
...
+        return lsp#utils#count_utf16_code_units(a:str)
+    endif
+    return strlen(a:str)
+endfunction
+
+function! lsp#utils#col(m) abort
+    if g:lsp_use_utf16
+      return lsp#utils#strlen(getline(a:m)[:col(a:m)])
+    endif
+    return a:m
```
Comment (hrsh7th) [inline autoload/lsp/utils.vim]: I met strange vim's behavior at trying to use this branch.

The behavior is `col(a:m)` is not match `lsp#utils#strlen(getline(a:m)[:col(a:m)])` in insert-mode.

In my environment, those unmatching were fixable below code.

```
function! lsp#utils#col(m) abort
    if g:lsp_use_utf16
        let col = lsp#utils#strlen(getline(a:m)[:col(a:m) - 1])

        if mode() ==# 'i' && col(a:m) == col('$')
            let col = col + 1
        endif

        return col
    endif
    return col(a:m)
endfunction
```

(This branch is very great work for me. Thanks mattn)

**Comment by clason** [comment]
@mattn Unfortunately, this PR doesn't help with the simple example from #425 (not even with @hrsh7th change). Am I missing something?

**Comment by hrsh7th** [comment]
@clason I missunderstood. My environemnt had other patch. It solves utf8 character problems maybe (This patch is `vim-lsc`'s diff logic). But still remaining utf16 character problem after applying this patch.
I'm sorry, I should had learning character encoding and checking my environment before post comment...

**Comment by mattn** [comment]
If someone already found the problem of this patch, please point it. Or if you already have another patch, please send new PR.

**Comment by clason** [comment]
@hrsh7th Could you share your "other patch"? If it works for UTF-8, that'd already be progress.

**Comment by hrsh7th** [comment]
I opend PR (#447).

It includes `other patch` (it is ported vim-lsc's diff logic).

@clason Please test #447.

Sorry for my confusing movements...

**Comment by mattn** [comment]
Fixed by https://github.com/prabirshrestha/vim-lsp/pull/447

**Pull request closed without merging**